pkg/skaffold/build/jib/jib_sync.go

```diff
@@ -0,0 +1,213 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jib
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"time"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/filemon"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+	"github.com/pkg/errors"
+)
+
+var syncLists = make(map[string]SyncMap)
+
+type SyncMap struct {
+	Direct    []SyncEntry `json:"direct"`
+	Generated []SyncEntry `json:"generated"`
+}
+
+type SyncEntry struct {
+	Src  string `json:"src"`
+	Dest string `json:"dest"`
+
+	filetime time.Time
+}
+
+func InitSync(ctx context.Context, workspace string, artifact *latest.JibArtifact) error {
+	syncMap, err := getSyncMap(ctx, workspace, artifact)
+	if (err != nil) {
+		return err
+	}
+	syncLists[artifact.Project] = *syncMap
+	return nil
+}
+
+// returns toCopy, toDelete, error
+func GetSyncDiff(ctx context.Context, workspace string, artifact *latest.JibArtifact, e filemon.Events) (map[string][]string, map[string][]string, error) {
+
+	// if anything that was modified was a buildfile, do NOT sync, do a rebuild
+	buildFiles := GetBuildDefinitions(artifact)
+	for _, f := range e.Modified {
+		if !filepath.IsAbs(f) {
+			if ff, err := filepath.Abs(f); err != nil{
+				return nil, nil, err
+			} else {
+				f = ff
+			}
+		}
+		for _, bf := range buildFiles {
+			if f == bf {
+				return nil, nil, nil
+			}
+		}
+	}
+
+	// it seems less than efficient to keep the original JSON structure when doing look ups, so maybe we should only use the json objects for serialization
+	// and store the sync data in a better type?
+	oldSyncMap := syncLists[artifact.Project]
+
+
+	// if all files are modified and direct, we don't need to build anything
+	if len(e.Deleted) == 0 || len(e.Added) == 0 {
+		matches := make(map[string][]string)
+		for _, f := range e.Modified {
+			for _, se := range oldSyncMap.Direct {
+				// filemon.Events doesn't seem to make any guarantee about the paths,
+				// so convert them to absolute paths (that's what jib provides)
+				if !filepath.IsAbs(f) {
+					if ff, err := filepath.Abs(f); err != nil{
+						return nil, nil, err
+					} else {
+						f = ff
+					}
+				}
+				if se.Src == f {
+					matches[se.Src] = []string{se.Dest}
+					break
+				}
+			}
+		}
+		if len(matches) == len(e.Modified) {
+			return matches, nil, nil
+		}
+	}
+
+	if len(e.Deleted) != 0 {
+		// change into logging
+		fmt.Println("Deletions are not supported by jib auto sync at the moment")
+		return nil, nil, nil;
+	}
+
+	// we need to do another build and get a new sync map
+	newSyncMap, err := getSyncMap(ctx, workspace, artifact)
```

Comment:
Oh, so `getSyncMap` is really supposed to trigger a build itself (i.e., doing `package`)?

I'd like to understand how this could potentially interfere (race condition) with normal file watching, as we talked about before.